File: sortega/python/calculator.py
"""Calculator code kata"""

import re

def add(input_text):
    """Adds comma-separated numbers."""
    if input_text == None:
        return 0

    lines = input_text.splitlines()

    # custom delimiter?
    if lines and lines[0].startswith('//'):
        delimiter_spec = lines.pop(0)[2:]
        delimiters = [multi or simple 
            for (multi, simple)
            in re.findall(r'\[(.*?)\]|(.)', delimiter_spec)]
    else:
        delimiters = [',']

    del_regex = '|'.join([re.escape(delimiter) for delimiter in delimiters])
    tokens = sum([re.split(del_regex, line) for line in lines], [])
    numbers = [int(token) for token in tokens]
    if [n for n in numbers if n < 0]:
        raise ValueError("Negatives not allowed")

    return sum([n for n in numbers if n <= 1000], 0)

File: sortega/ruby/calculator.rb
def add(numbers)
  return 0 unless numbers != nil
  lines = numbers.split(/\n/)

  if lines[0] =~ /^\/\//
    delimiter_spec = lines.shift[2..-1]
    delimiters = delimiter_spec.split(/(\[|\])+/)
    delimiters.select { |e| e }
  else
    delimiters = [ ',' ]
  end
  del_regex = (delimiters.map {|del| Regexp.escape(del) }).join('|')

  total = 0
  lines.each do |line|
    line.split(/#{del_regex}/).each do |n|
      value = n.to_i
      if value < 0
        raise ArgumentError.new('negatives not allowed')
      elsif value <= 1000
        total += value
      end
    end
  end

  return total
end
